File: apps/identity/src/account/exceptions/exceptions.ts
import { ExceptionBase } from '@libs/exceptions';
import { EMAIL_USED, INVALID_EMAIL_PROVIDER, USERNAME_USED } from './exception.codes';

export class EmailAlreadyInUseException extends ExceptionBase {
  constructor(email: string) {
    super(`Email '${email}' is already in use.`, EMAIL_USED);
  }
}

export class UsernameAlreadyInUseException extends ExceptionBase {
  constructor(username: string) {
    super(`Username '${username}' is already in use.`, USERNAME_USED);
  }
}

export class InvalidEmailProviderException extends ExceptionBase {
  constructor(provider: string) {
    super(`Email provider '${provider}' is not valid email provider.`, INVALID_EMAIL_PROVIDER);
  }
}

File: src/Modules/Identity/Thrive.Modules.Identity.Application/Exceptions/Exceptions.cs
using Thrive.Shared.Abstractions.Exceptions;
using Thrive.Shared.Application.Exceptions;

namespace Thrive.Modules.Identity.Application.Exceptions;

internal sealed class EmailAlreadyUsedException(string email)
    : BaseException($"Email '{email}' is already in use", ExceptionCodes.EmailAlreadyUsed);

internal sealed class UsernameAlreadyUsedException(string username)
    : BaseException($"Username '{username}' is already in use", ExceptionCodes.UsernameAlreadyUsed);

internal sealed class InvalidEmailProviderException(string emailProvider)
    : BaseException($"'{emailProvider}' is not valid email provider.", ExceptionCodes.InvalidEmailProvider);

internal sealed class InvalidCredentialsException() : UnauthorizedException("Invalid credentials");